poppy/zernike.py

```diff
@@ -702,17 +702,42 @@ def arbitrary_basis(aperture, nterms=15, rho=None, theta=None):
     # code submitted by Arthur Vigan - see https://github.com/mperrin/poppy/issues/166
 
     shape = aperture.shape
-    npix  = shape[0]
-
-    A = aperture.sum()
+    assert len(shape) == 2 and shape[0] == shape[1], \
+        "only square aperture arrays are supported"
+
+    # To avoid clipping the aperture, we precompute the zernike modes
+    # on an array oversized s.t. the zernike disk circumscribes the 
+    # entire aperture. We then slice the zernike array down to the 
+    # requested array size and cut the aperture out of it.
+
+    # get max extent of aperture from array center
+    yind, xind = np.where(aperture > 0)
+    distance = np.sqrt( (yind-shape[0]/2.)**2 + (xind-shape[1]/2.)**2 )
+    max_extent = distance.max()
+
+    # calculate padding for oversizing zernike_basis
+    ceil = lambda x: np.ceil(x) if x > 0 else 0 # avoid negative values
+    padding = ( int(ceil((max_extent - shape[0] / 2.))),
+                int(ceil((max_extent - shape[1] / 2.))) )
+    padded_shape = (shape[0] + padding[0] * 2, shape[1] + padding[1] * 2)
+    npix = padded_shape[0]
+
+    # pad theta and rho arrays
+    if theta is not None:
+        theta = np.pad(theta,padding,mode='constant',constant_values=0.)
+    if rho is not None:
+        rho = np.pad(rho,padding,mode='constant',constant_values=1.)
```

Comment:
I agree this handling of `theta` and `rho` doesn't feel right. As I recall the intent of this code is a very flexible capability for users to hand in arbitary indices to specific pixels without actually having a full contiguous array (consider a case of a sparse pupil which is mostly zeros; it would be more efficient to only evaluate zernike polynomials for the subset of pixels with nonzero intensity.)

Note that if `rho` and `theta` are not None, the `npix` argument is completed ignored inside `zernike_basis`.  Probably what I'd do here myself is check whether these arguments are present, and if so, just skip all the padding steps entirely. It's an advanced mode so it seems safe to declare that users who are using those arguments ought to be responsible for their own aperture padding.